File: production/client/src/main/scala/kpn/client/components/subset/nodes/SubsetOrphanNodeFilter.scala
package kpn.client.components.subset.nodes

import japgolly.scalajs.react.CallbackTo
import kpn.client.filter.FilterOptions
import kpn.client.filter.Filters
import kpn.client.filter.TimeFilterKind
import kpn.client.filter.TimestampFilter
import kpn.shared.NodeInfo
import kpn.shared.TimeInfo

class SubsetOrphanNodeFilter(timeInfo: TimeInfo, criteria: SubsetOrphanNodeFilterCriteria, updateCriteria: (SubsetOrphanNodeFilterCriteria) => Unit) {

  private val lastUpdatedFilter = new TimestampFilter[NodeInfo](
    criteria.lastUpdated,
    _.lastUpdated,
    timeInfo,
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.ALL))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.LAST_WEEK))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.LAST_MONTH))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.LAST_YEAR))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.OLDER))
    }
  )

  private val allFilters = new Filters[NodeInfo](
    lastUpdatedFilter
  )

  def filter(nodes: Seq[NodeInfo]): Seq[NodeInfo] = {
    nodes.filter(allFilters.passes)
  }

  def filterOptions(nodes: Seq[NodeInfo]): FilterOptions = {

    val totalCount = nodes.size
    val filteredCount = nodes.count(allFilters.passes)

    val lastUpdated = lastUpdatedFilter.filterOptions(allFilters, nodes)

    val groups = Seq(
      lastUpdated
    ).flatten

    FilterOptions(
      filteredCount,
      totalCount,
      groups: _*
    )
  }
}

File: frontend/src/app/analysis/subset/internal/orphan-nodes/components/subset-orphan-node-filter.ts
import { WritableSignal } from '@angular/core';
import { OrphanNodeInfo } from '@api/common/orphan-node-info';
import { TimeInfo } from '@api/common/time-info';
import { FilterOptions } from '@app/shared/kpn/filter/filter-options';
import { Filters } from '@app/shared/kpn/filter/filters';
import { TimestampFilter } from '@app/shared/kpn/filter/timestamp-filter';
import { TimestampFilterKind } from '@app/shared/kpn/filter/timestamp-filter-kind';
import { SubsetOrphanNodeFilterCriteria } from './subset-orphan-node-filter-criteria';

export class SubsetOrphanNodeFilter {
  private readonly lastUpdatedFilter = new TimestampFilter<OrphanNodeInfo>(
    this.criteria().lastUpdated,
    (row) => row.lastUpdated,
    this.timeInfo,
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.all })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.lastWeek })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.lastMonth })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.lastYear })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.older }))
  );

  private readonly allFilters = new Filters<OrphanNodeInfo>(this.lastUpdatedFilter);

  constructor(
    private readonly criteria: WritableSignal<SubsetOrphanNodeFilterCriteria>,
    private readonly timeInfo: TimeInfo
  ) {}

  filter(nodes: ReadonlyArray<OrphanNodeInfo>): ReadonlyArray<OrphanNodeInfo> {
    return nodes.filter((node) => this.allFilters.passes(node));
  }

  filterOptions(nodes: ReadonlyArray<OrphanNodeInfo>): FilterOptions {
    const totalCount = nodes.length;
    const filteredCount = nodes.filter((node) => this.allFilters.passes(node)).length;
    const lastUpdated = this.lastUpdatedFilter.filterOptions(this.allFilters, nodes);
    const groups = [lastUpdated].filter((g) => g !== null);
    return new FilterOptions(filteredCount, totalCount, groups);
  }
}
